lib/msprime.c

```diff
@@ -2818,17 +2818,22 @@ msp_run_dtwf(msp_t *self, double max_time, unsigned long max_events)
                 }
             }
         }
+        free(node_trees);
+        node_trees = NULL;
 
-        if (self->next_demographic_event != NULL) {
-            if (self->next_demographic_event->time <= self->time) {
-                ret = msp_apply_demographic_events(self);
-                if (ret != 0) {
-                    goto out;
-                }
+        /* Demographic events set the simulation time to the time of the event.
+         * In the DTWF, this would prevent more than one event occurring per
+         * generation, and throw off the time between generations. We avoid
+         * this by saving the current time and returning to it. */
+        cur_time = self->time;
+        while (self->next_demographic_event != NULL &&
+                self->next_demographic_event->time <= self->time) {
+            ret = msp_apply_demographic_events(self);
+            if (ret != 0) {
+                goto out;
             }
+            self->time = cur_time;
```

Comment:
Should't resetting time happen *outside* the loop? It's a bit confusing to see it being done in there.

Reply:
True, but if there were two events between generations, say at times 0.5 and 0.6, then the first would set the time to 0.5 and the next wouldn't execute. This makes sure we get them all.

Reply:
This is pretty confusing ... so self->time will be 1.0 when we enter this loop, then when we apply the event it'll get reset back to 0.5, and then we don't run it because self->time < 0.6.

What if we did it like:
```C
                cur_time = self->time;
￼	        while (self->next_demographic_event != NULL &&
￼	                self->next_demographic_event->time <= cur_time) {
￼	            ret = msp_apply_demographic_events(self);
￼	            if (ret != 0) {
￼	                goto out;
￼	            }
            }
￼	    self->time = cur_time;
```
That seems less confusing to me?

Reply:
Yep that looks better!